src/dlsnode/storage/BucketFile.d

```diff
@@ -334,6 +343,70 @@ public class BucketFile: OutputStream
         }
     }
 
+
+    /**************************************************************************
+
+        Opens the bucket file and reads or writes the bucket header in a
+        non-blocking fashion.
+
+        Params:
+            path = file path
+            suspended_job = JobNotification to block
+                the fiber on until read is completed
+            file_buffer = buffer used for buffered input
+            style = style in which file will be open for
+
+    **************************************************************************/
+
+    public bool openAsync (cstring path,
+            JobNotification suspended_job,
+            void[] file_buffer,
+            File.Style style = File.ReadExisting)
+    in
+    {
+        assert(style == File.ReadExisting || style == File.ReadWriteAppending,
+                "Currently BucketFile only supports ReadExisting or ReadWriteAppending");
+    }
+    body
+    {
+        if (!this.is_open_ || style == File.ReadWriteAppending)
+        {
+            this.promise.reset();
+            this.buffered_input.reset(file_buffer);
+            this.file.open(path, style);
+            this.file_length_ = this.file.length;
+            this.file_pos_ = this.file.position;
+            this.is_open_ = true;
+        }
+
+        // In case we opened this file for writting,
+        // but the position is at 0 (empty file), we need to
+        // write the magic header.
+        if (style == File.ReadWriteAppending && this.file_length_ == 0)
+        {
+            this.writeBucketHeader();
+            return true;
+        }
+        else
+        {
+            // in case we're openning already existing file,
+            // just read the existing header (if any).
+            // NOTE: there's no need to reposition cursor. The reason is that
+            // we have opened underlying file with O_APPEND and writes will
+            // always be appended to the end of the file
+            auto open = this.readBucketHeaderAsync(suspended_job);
+
+            // if the file is open for appending, seek to the end
```

Comment:
Hmm, this seems to contradict the NOTE above. I'm probably missing something though :)

Reply:
No, you're not missing anything, the top comment is outdated - because we do seek to the beginning of the file to read the bucket header and then we need to seek back to the end of the file.

Reply:
Ah cool. It's difficult to say with this code, as I'm not that familiar with it ;)